File: Project Netwerken/src/NodeJS/global/model/command/Command.js
const varint = require('../../utils/varint/varint');

class Command {
    constructor(options) {
        this.type = options.type;
        if (options.buffer) {
            // constructor(type, buffer)
            this.nodeId = varint.decode(options.buffer);
        } else {
            // constructor(type, nodeId, vectorCock)
            this.nodeId = options.nodeId || options.authorId;
            this.vectorClock = options.vectorClock;
        }
        this.type = options.type;
        this.nodeId = options.nodeId || varint.decode(options.buffer);
        if (options.vectorClock) this.vectorClock = options.vectorClock;
    }

    /*	// OLD CONSTRUCTOR
        constructor(type, buffer){
            this.type = type;
            this.nodeId = varint.decode(buffer);
        }

        // OLD CONSTRUCTOR
        constructor(type, nodeId, vectorClock) {
            if (new.target === Abstract) {
                throw new TypeError("Cannot construct Abstract instances directly");
            }
            this.type = type;
            this.nodeId = nodeId;
            this.vectorClock = vectorClock;
        }*/

    write(buffer) {
        buffer.write(varint.encode(this.type));
        buffer.write(varint.encode(this.nodeId));
        this.vectorClock.write(buffer);
    }
}

module.exports = Command;

File: Project Netwerken/src/Global/src/model/command/Command.java
package model.command;

import utils.VarInt;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * Commands that will be encrypted in the DataPacket. This is where the data is stored.
 */
public abstract class Command {
    private int type;
    private int nodeId; // Used for vector clock
    private VectorClock vectorClock;

    public Command(int type, ByteBuffer buffer){
        this.type = type;
        nodeId = VarInt.readUnsignedVarInt(buffer);
        vectorClock = new VectorClock(buffer);
    }

    /**
     * @param nodeId vectorclock index
     * @param vectorClock the new vectorclock
     */
    public Command(int type, int nodeId, VectorClock vectorClock) {
        assert vectorClock != null : "Vector clock is null.";
        this.type = type;
        this.nodeId = nodeId;
        this.vectorClock = vectorClock;
    }

    public int getAuthorId() {
        return nodeId;
    }

    public VectorClock getVectorClock() {
        return vectorClock;
    }

    public int getType() {
        return type;
    }
    /**
     * This method will translate the whole command to byte[].
     * @return byte[]
     */
    public void write(ByteArrayOutputStream out) throws IOException {
        assert out != null : "ByteArrayOutputStream is null.";
        out.write(VarInt.writeUnsignedVarInt(type));
        out.write(VarInt.writeUnsignedVarInt(nodeId));
        vectorClock.write(out);
    }
}
